src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,223 @@
+        [input_data = input.begin<cudf::timestamp_D>(), fixed_transitions, tz_index] __device__ (auto const i) {
+            auto const timestamp = input_data[i];
+            return convert_timestamp_timezone<cudf::timestamp_D>(timestamp, fixed_transitions, tz_index, true);
+          }
+      );
+      break;
+    case cudf::type_id::TIMESTAMP_SECONDS: 
+      thrust::transform(rmm::exec_policy(stream),
+        thrust::make_counting_iterator(0),
+        thrust::make_counting_iterator(num_rows),
+        results->mutable_view().begin<cudf::timestamp_s>(),
+        [input_data = input.begin<cudf::timestamp_s>(), fixed_transitions, tz_index] __device__ (auto const i) {
+            auto const timestamp = input_data[i];
+            return convert_timestamp_timezone<cudf::timestamp_s>(timestamp, fixed_transitions, tz_index, true);
+          }
+      );
+      break;
+    case cudf::type_id::TIMESTAMP_MILLISECONDS: 
+      thrust::transform(rmm::exec_policy(stream),
+        thrust::make_counting_iterator(0),
+        thrust::make_counting_iterator(num_rows),
+        results->mutable_view().begin<cudf::timestamp_ms>(),
+        [input_data = input.begin<cudf::timestamp_ms>(), fixed_transitions, tz_index] __device__ (auto const i) {
+            auto const timestamp = input_data[i];
+            return convert_timestamp_timezone<cudf::timestamp_ms>(timestamp, fixed_transitions, tz_index, true);
+          }
+      );
+      break;
+    case cudf::type_id::TIMESTAMP_MICROSECONDS: 
+      thrust::transform(rmm::exec_policy(stream),
+        thrust::make_counting_iterator(0),
+        thrust::make_counting_iterator(num_rows),
+        results->mutable_view().begin<cudf::timestamp_us>(),
+        [input_data = input.begin<cudf::timestamp_us>(), fixed_transitions, tz_index] __device__ (auto const i) {
+            auto const timestamp = input_data[i];
+            return convert_timestamp_timezone<cudf::timestamp_us>(timestamp, fixed_transitions, tz_index, true);
+          }
+      );
+      break;
+    default: 
+      CUDF_FAIL("Unsupported timestamp unit for timezone conversion");
+  }
+
+
+  return results;
+}
+
+std::unique_ptr<column> convert_utc_timestamp_to_timezone(cudf::column_view const& input,
+  cudf::table_view const& transitions,
+  size_type tz_index,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr) {
+
+  auto const type = input.type().id();
+  auto num_rows = input.size();
+
+  // get the fixed transitions
+  auto const ft_cdv_ptr = column_device_view::create(transitions.column(0), stream);
+  lists_column_device_view fixed_transitions = cudf::detail::lists_column_device_view{*ft_cdv_ptr};
+
+  auto results = cudf::make_timestamp_column(input.type(), input.size(),
+                                            cudf::detail::copy_bitmask(input, stream, mr),
+                                            input.null_count(), stream, mr);
+
+  switch (type) {
+    case cudf::type_id::TIMESTAMP_DAYS: 
```

Comment:
Do we really need anything here except microseconds?  After all that is all that Spark supports. If we think there are more that we could/should support, then fine, but DAYS in particular I don't think should be supported. There are no tests for it, and I don't think Spark will do any conversion on Date/DAYS to different timezones.

Reply:
Actually I was planning to remove DAYS, so I will remove it. I do think that  Spark can cast timestamp to a date and vice-versa (timezone aware), but I'm not sure that this code would handle it anyways.